### PR TITLE
Order book reserve

### PR DESCRIPTION
**Pull request by ilanDoron**
not for pull. only for code review.

### CODE REVIEW

File: contracts/KyberNetwork.sol [hunk 1]
```diff
@@ -91,21 +125,30 @@ contract KyberNetwork is Withdrawable, Utils2, KyberNetworkInterface {
...
-    function addReserve(KyberReserveInterface reserve, bool add) public onlyAdmin {
+    function addReserve(KyberReserveInterface reserve, bool add, bool isPermissionless) public onlyOperator {
 
         if (add) {
-            require(!isReserve[reserve]);
+            require(reserveType[reserve] == ReserveType.NONE);
             reserves.push(reserve);
-            isReserve[reserve] = true;
+
+            if (isPermissionless) {
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: cto likes the ternary operator.
= isPermissionless ? A : B

File: contracts/KyberNetwork.sol [hunk 2]
```diff
@@ -226,9 +271,18 @@ contract KyberNetwork is Withdrawable, Utils2, KyberNetworkInterface {
     }
 
     function getUserCapInWei(address user) public view returns(uint) {
+        if (whiteListContract == address(0)) return (2 ** 255);
         return whiteListContract.getUserCapInWei(user);
     }
 
+    function verifyUserCap(address trader, uint weiValue) internal view returns(bool) {
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: I don't think we need this function

File: contracts/KyberNetwork.sol [hunk 3]
```diff
@@ -264,11 +318,32 @@ contract KyberNetwork is Withdrawable, Utils2, KyberNetworkInterface {
         return infoFields[field];
     }
 
+    function searchBestRate(ERC20 src, ERC20 dest, uint srcAmount) public view returns(address, uint) {
+        return searchBestRatePermissioned(src, dest, srcAmount, true);
+    }
+
+    function reservesPerTokenSrc(address token, uint index) public view returns(address) {
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: why do we need this function?

File: contracts/KyberNetwork.sol [hunk 4]
```diff
@@ -264,11 +318,32 @@ contract KyberNetwork is Withdrawable, Utils2, KyberNetworkInterface {
...
+        if (index >= reservesTokenSrc[token].length) return (address(0));
+
+        return reservesTokenSrc[token][index];
+    }
+
+    function reservesPerTokenDest(address token, uint index) public view returns(address) {
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: also this function. why?

File: contracts/KyberNetwork.sol [hunk 5]
```diff
@@ -311,35 +389,48 @@ contract KyberNetwork is Withdrawable, Utils2, KyberNetworkInterface {
 
             if (numRelevantReserves > 1) {
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: should be >= 1, and else return 0.

File: contracts/KyberNetwork.sol [hunk 6]
```diff
@@ -311,35 +389,48 @@ contract KyberNetwork is Withdrawable, Utils2, KyberNetworkInterface {
...
         return (reserveArr[bestReserve], bestRate);
     }
     /* solhint-enable code-complexity */
 
-    function findBestRateTokenToToken(ERC20 src, ERC20 dest, uint srcAmount) internal view
+    function findBestRateTokenToToken(ERC20 src, ERC20 dest, uint srcAmount, bytes hint) internal view
         returns(BestRateResult result)
     {
-        (result.reserve1, result.rateSrcToEth) = searchBestRate(src, ETH_TOKEN_ADDRESS, srcAmount);
+        bool usePermissionless = false;
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: as we discussed, default is true.

File: contracts/KyberNetwork.sol [hunk 7]
```diff
@@ -311,35 +389,48 @@ contract KyberNetwork is Withdrawable, Utils2, KyberNetworkInterface {
...
+
+        // PERM ascii == P = 80, E = 69, R = 82, M = 77
+        if ((hint.length >= 4) &&
+            (keccak256(hint[0], hint[1], hint[2], hint[3]) == keccak256(byte(80), byte(69), byte(82), byte(77)))){
+            usePermissionless = true;
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: should be set to false.
maybe change to ASCII "ILAN" :)

File: contracts/KyberNetwork.sol [hunk 8]
```diff
@@ -264,11 +318,32 @@ contract KyberNetwork is Withdrawable, Utils2, KyberNetworkInterface {
...
+    }
+
     /* solhint-disable code-complexity */
-    // Not sure how solhing defines complexity. Anyway, from our point of view, below code follows the required
-    //  algorithm to choose a reserve, it has been tested, reviewed and found to be clear enough.
+    // Regarding complexity. Below code follows the required algorithm for choosing a reserve.
+    //  It has been tested, reviewed and found to be clear enough.
     //@dev this function always src or dest are ether. can't do token to token
-    function searchBestRate(ERC20 src, ERC20 dest, uint srcAmount) public view returns(address, uint) {
+    function searchBestRatePermissioned(ERC20 src, ERC20 dest, uint srcAmount, bool usePermissionless)
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: the name is bad. it suggest you only look in permissioned reserves.

File: contracts/KyberNetwork.sol [hunk 9]
```diff
@@ -390,17 +480,11 @@ contract KyberNetwork is Withdrawable, Utils2, KyberNetworkInterface {
...
-            if (tradeInput.src == ETH_TOKEN_ADDRESS) {
-                tradeInput.trader.transfer(tradeInput.srcAmount - actualSrcAmount);
-            } else {
-                tradeInput.src.transfer(tradeInput.trader, (tradeInput.srcAmount - actualSrcAmount));
-            }
-        }
-
         // verify trade size is smaller than user cap
-        require(weiAmount <= getUserCapInWei(tradeInput.trader));
+        require(verifyUserCap(tradeInput.trader, weiAmount));
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: why did you change the code here?

File: contracts/KyberNetwork.sol [hunk 10]
```diff
@@ -390,17 +480,11 @@ contract KyberNetwork is Withdrawable, Utils2, KyberNetworkInterface {
...
+
+        //if any "change" in src token, send back to trader
+        handleChange(tradeInput.src, tradeInput.srcAmount, actualSrcAmount, tradeInput.trader);
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: should check return value.

File: contracts/KyberNetwork.sol [hunk 11]
```diff
@@ -504,6 +587,19 @@ contract KyberNetwork is Withdrawable, Utils2, KyberNetworkInterface {
         return true;
     }
 
+    /// when user sets max dest amount we could have too many source tokens == change. so we send it back to user.
+    function handleChange (ERC20 src, uint srcAmount, uint requiredSrcAmount, address trader) internal {
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: please return true.

File: contracts/KyberNetworkInterface.sol [hunk 12]
```diff
@@ -16,5 +16,5 @@ interface KyberNetworkInterface {
         returns (uint expectedRate, uint slippageRate);
 
     function tradeWithHint(address trader, ERC20 src, uint srcAmount, ERC20 dest, address destAddress,
-        uint maxDestAmount, uint minConversionRate, address walletId, bytes hint) public payable returns(uint);
+        uint maxDestAmount, uint minConversionRate, address walletId, bytes hint) external payable returns(uint);
```
Comment (yaronvel) [inline contracts/KyberNetworkInterface.sol]: I am not sure we can change network interface. it changes the binary of the proxy. need to discuss it.

File: contracts/permissionless/OrdersFactory.sol [hunk 13]
```diff
@@ -0,0 +1,16 @@
+pragma solidity ^0.4.0;
```
Comment (yaronvel) [inline contracts/permissionless/OrdersFactory.sol]: 4.0?!!!

File: contracts/permissionless/PermissionlessOrderBookReserveLister.sol [hunk 14]
```diff
@@ -0,0 +1,89 @@
+pragma solidity 0.4.18;
+
+
+import "./OrderBookReserve.sol";
+import "../FeeBurnerInterface.sol";
+
+
+contract KyberNetworkInterface {
```
Comment (yaronvel) [inline contracts/permissionless/PermissionlessOrderBookReserveLister.sol]: I don't think you should have it here.

File: contracts/KyberNetworkInterface.sol (hunk 12 above)
Comment (ilanDoron) [reply contracts/KyberNetworkInterface.sol]: we have to do it to use the reentrant code.
as u suggest in your documentation...

File: contracts/permissionless/PermissionlessOrderBookReserveLister.sol (hunk 14 above)
Comment (ilanDoron) [reply contracts/permissionless/PermissionlessOrderBookReserveLister.sol]: well
all the below are not defined in the interface.
I can add them there.
but will require changes in contract.
Since if interface is inherited the "automatic"  API for public parameters should be modified.

File: contracts/KyberNetwork.sol [hunk 15]
```diff
@@ -11,22 +11,56 @@ import "./ExpectedRateInterface.sol";
...
     FeeBurnerInterface    public feeBurnerContract;
     address               public kyberNetworkProxyContract;
+
     uint                  public maxGasPriceValue = 50 * 1000 * 1000 * 1000; // 50 gwei
     bool                  public isEnabled = false; // network is enabled
     mapping(bytes32=>uint) public infoFields; // this is only a UI field for external app.
-    mapping(address=>address[]) public reservesPerTokenSrc; //reserves supporting token to eth
-    mapping(address=>address[]) public reservesPerTokenDest;//reserves support eth to token
+
+    mapping(address=>address[]) public reservesTokenSrc; //reserves supporting token to eth
```
[the remark this answers is not in the document]
Comment (ilanDoron) [reply contracts/KyberNetwork.sol]: this change reverted

**Comment by ilanDoron** [comment]
PR will be added again after all required code changes
